### PR TITLE
Exception Handling Part 1

### PR DESCRIPTION
**Pull request by andyday**
This is part 1 of a 2 part feature.  The goal is to provide conversion of Exceptions to proper HTTP Responses.  This PR sets the stage by allowing for the registering of an ExceptionHandler with the fall back to a DefaultExceptionHandler that does nothing more than re-throw the exception.  In part 2 the DefaultExceptionHandler will look for specific types of exceptions and return appropriate Responses based on a reasonable set of rules...

### CODE REVIEW

File: xrpc/src/main/java/com/nordstrom/xrpc/server/ExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.nordstrom.xrpc.server;
+
+import io.netty.handler.codec.http.HttpResponse;
+
+/** A Handler for mapping exceptions to Http Responses. */
+@FunctionalInterface
+public interface ExceptionHandler {
+  HttpResponse handle(XrpcRequest request, Exception exception) throws Exception;
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/ExceptionHandler.java]: It's not clear under what conditions this should throw an exception. Document? Ideally, this should _never_ throw exceptions (it's meant as a fallback).

File: xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java [hunk 2]
```diff
@@ -109,6 +109,10 @@ private void addResponseCodeMeters(XrpcConnectionContext.Builder contextBuilder)
     }
   }
 
+  public void registerExceptionHandler(ExceptionHandler handler) {
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: Document? In particular, it's unclear if this overwrites the only exception handler, or appends an exception handler to a list.

File: xrpc/src/main/java/com/nordstrom/xrpc/server/UrlRouter.java [hunk 3]
```diff
@@ -54,12 +53,16 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
...
 
-      ObjectMapper mapper = xctx.getMapper();
       XrpcRequest xrpcRequest =
           new XrpcRequest(request, xctx.getMapper(), match.getGroups(), ctx.channel());
       xrpcRequest.setData(request.content());
 
-      HttpResponse resp = match.getHandler().handle(xrpcRequest);
+      HttpResponse resp;
+      try {
+        resp = match.getHandler().handle(xrpcRequest);
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/UrlRouter.java]: You should probably scope the `try` block to only the `handle` call. We don't want errors in our `getHandler()` method to be sent to the user.

File: xrpc/src/main/java/com/nordstrom/xrpc/server/DefaultExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package com.nordstrom.xrpc.server;
+
+import io.netty.handler.codec.http.HttpResponse;
+
+public class DefaultExceptionHandler implements ExceptionHandler {
+  @Override
+  public HttpResponse handle(XrpcRequest request, Exception exception) throws Exception {
+    // TODO (AD): Use configured response content/type and response specific exceptions to default
+    // to meaningful responses here
+    throw exception;
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/DefaultExceptionHandler.java]: I would omit the default handler if all it's doing is rethrowing, since that can be implemented in code (`if handler == null throw e`).

File: xrpc/src/main/java/com/nordstrom/xrpc/server/UrlRouter.java [hunk 5]
```diff
@@ -54,12 +53,16 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
...
       CompiledRoutes.Match match = xctx.getRoutes().match(path, request.method());
 
-      ObjectMapper mapper = xctx.getMapper();
       XrpcRequest xrpcRequest =
           new XrpcRequest(request, xctx.getMapper(), match.getGroups(), ctx.channel());
       xrpcRequest.setData(request.content());
 
-      HttpResponse resp = match.getHandler().handle(xrpcRequest);
+      HttpResponse resp;
+      try {
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/UrlRouter.java]: `Handler.handle` is invoked twice in `Http2Handler` as well.

File: xrpc/src/main/java/com/nordstrom/xrpc/server/DefaultExceptionHandler.java (hunk 4 above)
Comment (andyday) [reply xrpc/src/main/java/com/nordstrom/xrpc/server/DefaultExceptionHandler.java]: at this point it is a placeholder.  in part 2 it will do more than just rethrow

**Comment by andyday** [comment]
@jkinkead i've moved the logic to compiled routes and added some clarifying documentation...

File: xrpc/src/main/java/com/nordstrom/xrpc/server/CompiledRoutes.java [hunk 6]
```diff
@@ -68,20 +68,26 @@ public CompiledRoutes(
...
-                // Should be impossible.
-                throw new IllegalStateException("unexpected exception", e);
+                try {
+                  return request.getConnectionContext().getExceptionHandler().handle(request, e);
+                } catch (IOException | RuntimeException ioe) {
+                  // From child handler.
+                  throw ioe;
+                } catch (Exception ee) {
+                  // Should be impossible.
+                  throw new IllegalStateException("unexpected exception", ee);
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/CompiledRoutes.java]: You've inverted the wrapping here; you should pass through only the expected child exceptions:
```java
try {
  return timer.time(() -> userHandler.handle(request));
} catch (IOException | RuntimeException handlerE) {
  try {
    return request.getConnectionContext().getExceptionHandler().handle(request, handlerE);
  } catch (Exception ee) {
    // TODO: Handle gracefully. This should probably log and return a generic 500.
  }
} catch (Exception e) {
  // Should be impossible.
  throw new IllegalStateException("unexpected exception", ee);
}
```

File: xrpc/src/main/java/com/nordstrom/xrpc/server/ExceptionHandler.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+import io.netty.handler.codec.http.HttpResponse;
+
+/** A Handler for mapping exceptions to Http Responses. */
+@FunctionalInterface
+public interface ExceptionHandler {
+  /**
+   * Handle exception. This is used to handle an exception thrown by the registered Route Handler.
+   * It converts the Exception to an appropriate HttpResponse.
+   *
+   * @param request Request for which this Exception was generated
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/ExceptionHandler.java]: Lower-case first word of non-full-sentence Javadoc for `@param`.

File: xrpc/src/main/java/com/nordstrom/xrpc/server/ExceptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+   * @param exception Thrown exception
+   * @return HttpResponse appropriate for this Exception.
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/ExceptionHandler.java]: Omit terminal period for sentence fragment `@return`

File: xrpc/src/main/java/com/nordstrom/xrpc/server/ExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+   * @throws Exception Exception for any unhandled Exception passed in or during the conversion
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/ExceptionHandler.java]: I don't think we want to allow this; I don't see this as a valid use-case. Handlers should be in charge of generating a valid response for all requests.

If this throws, we should be catching it and returning a generic 500 response.

File: xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java [hunk 10]
```diff
@@ -109,6 +109,16 @@ private void addResponseCodeMeters(XrpcConnectionContext.Builder contextBuilder)
     }
   }
 
+  /**
+   * Set the customer exception handler. This handler is called to handle any exceptions thrown from
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: "custom" instead of "customer"?

File: xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java [hunk 11]
```diff
@@ -109,6 +109,16 @@ private void addResponseCodeMeters(XrpcConnectionContext.Builder contextBuilder)
...
+   * route handlers. This replaces any other exception handlers previously set.
+   *
+   * @param handler Exception Handler
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: Omit this per google style guide? (omit `@param` if it's already covered by existing doc).

File: xrpc/src/main/java/com/nordstrom/xrpc/server/XrpcRequest.java [hunk 12]
```diff
@@ -44,15 +43,14 @@
...
-  private final ObjectMapper mapper;
-
   /** The request to handle. */
   @Getter private final FullHttpRequest h1Request;
 
   @Getter private final Http2Headers h2Headers;
   @Getter private final EventLoop eventLoop;
   @Getter private final Channel upstreamChannel;
   @Getter private final ByteBufAllocator alloc;
+  @Getter private final XrpcConnectionContext connectionContext;
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/XrpcRequest.java]: I _think_ we should make this package:
```java
import lombok.AccessLevel;
...
@Getter(AccessLevel.PACKAGE) private final XrpcConnectionContext connectionContext;
```
I don't believe that users will ever need to interact with this. It's currently safe for them to interact with, but it might not always remain so, and it's cluttering up the API at the moment.

File: xrpc/src/main/java/com/nordstrom/xrpc/server/XrpcRequest.java [hunk 13]
```diff
@@ -63,10 +61,13 @@
   private ByteBuf data;
 
   public XrpcRequest(
-      FullHttpRequest request, ObjectMapper mapper, Map<String, String> groups, Channel channel) {
+      FullHttpRequest request,
+      XrpcConnectionContext ctx,
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/XrpcRequest.java]: Match parameter to member name?

File: xrpc/src/main/java/com/nordstrom/xrpc/server/XrpcRequest.java [hunk 14]
```diff
@@ -76,13 +77,13 @@ public XrpcRequest(
 
   public XrpcRequest(
       Http2Headers headers,
-      ObjectMapper mapper,
+      XrpcConnectionContext ctx,
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/XrpcRequest.java]: Match parameter to member name?

File: xrpc/src/main/java/com/nordstrom/xrpc/server/XrpcRequest.java [hunk 15]
```diff
@@ -222,7 +223,7 @@ public FullHttpResponse badRequestJsonResponse(Object body) throws IOException {
   private ByteBuf bodyToByteBuf(Object body) throws IOException {
     ByteBuf buf = getAlloc().directBuffer();
     OutputStream stream = new ByteBufOutputStream(buf);
-    mapper.writeValue(stream, body);
+    this.connectionContext.getMapper().writeValue(stream, body);
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/XrpcRequest.java]: Remove `this` for consistency with the rest of the file?

File: xrpc/src/main/java/com/nordstrom/xrpc/server/CompiledRoutes.java (hunk 6 above)
Comment (andyday) [reply xrpc/src/main/java/com/nordstrom/xrpc/server/CompiledRoutes.java]: that's what I get for doing this during a design review

Comment (andyday) [reply xrpc/src/main/java/com/nordstrom/xrpc/server/CompiledRoutes.java]: actually i didn't invert it.  i first want to pass any Exception to the exception handler and if any exception comes from that I want to split into more granular exception handling

File: xrpc/src/main/java/com/nordstrom/xrpc/server/XrpcRequest.java (hunk 12 above)
Comment (andyday) [reply xrpc/src/main/java/com/nordstrom/xrpc/server/XrpcRequest.java]: this could probably be said for some of the other XrpcRequest getters as well.

**Comment by andyday** [comment]
Think I've addressed all your comments...

File: xrpc/src/main/java/com/nordstrom/xrpc/XrpcConstants.java [hunk 16]
```diff
@@ -49,6 +49,9 @@
           .writeBytes(
               "This response is being send due to too many requests being sent to the server"
                   .getBytes(DEFAULT_CHARSET));
+
+  public static final byte[] INTERNAL_SERVER_ERROR_RESONSE =
+      "Internal Server Error".getBytes(DEFAULT_CHARSET);
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/XrpcConstants.java]: This just makes line 50's response look too wordy . . . :)

File: xrpc/src/main/java/com/nordstrom/xrpc/server/DefaultExceptionHandler.java [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+import io.netty.handler.codec.http.HttpResponseStatus;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class DefaultExceptionHandler implements ExceptionHandler {
+  @Override
+  public HttpResponse handle(XrpcRequest request, Exception exception) {
+    // TODO (AD): Use configured response content/type and response specific exceptions to default
+    // to meaningful responses here
+    log.error("Handler Exception: {}", exception.toString());
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/DefaultExceptionHandler.java]: This should probably log the stack trace, for debuggability:
```java
log.error("Handler Exception:", exception);
```

File: xrpc/src/main/java/com/nordstrom/xrpc/XrpcConstants.java (hunk 16 above)
Comment (andyday) [reply xrpc/src/main/java/com/nordstrom/xrpc/XrpcConstants.java]: i shortened the one on line 50 as it is the outlier.   a similar comment is coming on your PR that changes that line.

File: xrpc/src/main/java/com/nordstrom/xrpc/server/DefaultExceptionHandler.java (hunk 17 above)
Comment (andyday) [reply xrpc/src/main/java/com/nordstrom/xrpc/server/DefaultExceptionHandler.java]: doh.   in C# toString() does the trick.   good catch.

File: xrpc/src/main/java/com/nordstrom/xrpc/XrpcConstants.java (hunk 16 above)
Comment (jkinkead) [reply xrpc/src/main/java/com/nordstrom/xrpc/XrpcConstants.java]: 👍
